### ISSUE TITLE
Veracode scan identified multiple security flaws

### ISSUE THREAD

**Issue by bchornii** [jstedfast/MailKit#1530]
**Describe the bug**
Veracode scan of the project with MailKit nuget package installed (v.3.6.0) identified **56** security flaws. Please review and provide comments on them. 

**Platform (please complete the following information):**
 - OS: Windows
 - .NET Runtime: .NET 6
 - .NET Framework: [e.g. .Net Core, .NET 4.5, UWP, etc.]
 - MailKit Version: 3.6.0.

[MailKit_Flaws.zip](https://github.com/jstedfast/MailKit/files/10955947/MailKit_Flaws.zip)

**Comment by jstedfast** [comment]
This document is pretty useless, to be honest. If you want me to look at these issues, I'll need you to provide me with something more useful (and please, no Office documents).

I can pretty much guarantee that all of the "Use of a Broken or RIsky Cryptographic Algorithm" items in the list are because of MD5 but that's a meaningless warning because I can't get rid of MD5 if I want to support older email features like Content-MD5.

**Comment by bchornii** [comment]
This excel spreadsheet contains issues we received on Veracode scan. If it's more convenient I can share PDF doc with those instead, would it work ?

**Comment by jstedfast** [comment]
Maybe? The main problem with the Excel spreadsheet is that it doesn't seem to provide any class names, line numbers, etc.

It only tells me that it thinks there's a problem in a .ctor() in mailkit.dll which is useless.

**Comment by jstedfast** [comment]
Also note that half or so of these issues are in bouncycastle.dll which is another nuget and not part of MimeKit or MailKit.

**Comment by bchornii** [comment]
> Also note that half or so of these issues are in bouncycastle.dll which is another nuget and not part of MimeKit or MailKit.

Ok, I will create a separate ticket for it

**Comment by bchornii** [comment]
If you know that most of those shall be treated as a false positives that's fine, please let me know. My security team is asking me for comments regarding those issues, to get the idea how they should be treated

**Comment by bchornii** [comment]
Our latest scan also determined that system.security.cryptography.pkcs v.6.0 which is also referenced by MimeKit 3.6 contains several security flaws:
- <html>
<body>


3821 |  Medium | Neutral | 316  Cleartext Storage of Sensitive Information in Memory | 3/13/2023 9:07 AM EDT | system.security.cryptography.pkcs.dll | void DecodeCore(ref Formats.Asn1.AsnValueReader /*8*/, System.Formats.Asn1.Asn1Tag, System.ReadOnlyMemory, ref SecretBagAsn /*8*/): 31% | 1 Path | Open | None
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
3820 |  Medium  | Likely | 327  Use of a Broken or Risky Cryptographic Algorithm | 3/13/2023 9:07 AM EDT | system.security.cryptography.pkcs.dll | System.Security.Cryptography.SymmetricAlgorithm OpenAlgorithm(string): 52% | 1 Path | Open | None
3819 |  Medium  | Likely | 327  Use of a Broken or Risky Cryptographic Algorithm | 3/13/2023 9:07 AM EDT | system.security.cryptography.pkcs.dll | System.Security.Cryptography.SymmetricAlgorithm OpenAlgorithm(string): 55% | 1 Path | Open | None
3818 |  Medium  | Likely | 327  Use of a Broken or Risky Cryptographic Algorithm | 3/13/2023 9:07 AM EDT | system.security.cryptography.pkcs.dll | System.Security.Cryptography.SymmetricAlgorithm OpenAlgorithm(string): 59% | 1 Path | Open | None
3815 |   Medium | Likely | 327  Use of a Broken or Risky Cryptographic Algorithm | 3/13/2023 9:07 AM EDT | system.security.cryptography.pkcs.dll | System.Security.Cryptography.RC2 CreateRC2(): 97% | 1 Path | Open | None
3814 |   Medium | Likely | 327  Use of a Broken or Risky Cryptographic Algorithm | 3/13/2023 9:07 AM EDT | system.security.cryptography.pkcs.dll | System.Security.Cryptography.SymmetricAlgorithm OpenCipher(Asn1.AlgorithmIdentifierAsn, System.Nullable, ref System.Span /*1*/): 97% | 1 Path | Open | None
3812 |  Medium  | Likely | 327  Use of a Broken or Risky Cryptographic Algorithm | 3/13/2023 9:07 AM EDT | system.security.cryptography.pkcs.dll | System.Security.Cryptography.SymmetricAlgorithm OpenCipher(Asn1.AlgorithmIdentifierAsn, System.Nullable, ref System.Span /*1*/): 47% | 1 Path | Open | None
3810 |  Medium  | Likely | 327  Use of a Broken or Risky Cryptographic Algorithm | 3/13/2023 9:07 AM EDT | system.security.cryptography.pkcs.dll | void InitiateEncryption(System.Security.Cryptography.PbeParameters, ref System.Security.Cryptography.SymmetricAlgorithm /*1*/, ref string /*1*/, ref string /*1*/, ref bool): 42% | 1 Path | Open | None
3809 |   Medium | Likely | 327  Use of a Broken or Risky Cryptographic Algorithm | 3/13/2023 9:07 AM EDT | system.security.cryptography.pkcs.dll | int Decrypt(ref Asn1.AlgorithmIdentifierAsn /*8*/, System.ReadOnlySpan, System.ReadOnlySpan, System.ReadOnlySpan, System.Span): 48% | 1 Path | Open | None
3822 |   Medium | Neutral | 316  Cleartext Storage of Sensitive Information in Memory | 3/13/2023 9:07 AM EDT | system.security.cryptography.pkcs.dll | byte[] EncodeBagValue(System.Security.Cryptography.Oid, ref System.ReadOnlyMemory /*1*/): 27% | 1 Path | Open | None
3807 |  Medium  | Likely | 327  Use of a Broken or Risky Cryptographic Algorithm | 3/13/2023 9:07 AM EDT | system.security.cryptography.pkcs.dll | int Decrypt(ref Asn1.AlgorithmIdentifierAsn /*8*/, System.ReadOnlySpan, System.ReadOnlySpan, System.ReadOnlySpan, System.Span): 50% | 1 Path | Open | None
3805 |  Medium  | Likely | 327  Use of a Broken or Risky Cryptographic Algorithm | 3/13/2023 9:07 AM EDT | system.security.cryptography.pkcs.dll | int Decrypt(ref Asn1.AlgorithmIdentifierAsn /*8*/, System.ReadOnlySpan, System.ReadOnlySpan, System.ReadOnlySpan, System.Span): 45% | 1 Path | Open | None
3817 |   Medium | V.Likely | 326  Inadequate Encryption Strength | 3/13/2023 9:07 AM EDT | system.security.cryptography.pkcs.dll | System.Security.Cryptography.SymmetricAlgorithm OpenAlgorithm(System.Security.Cryptography.Pkcs.AlgorithmIdentifier): 66% | 1 Path | Open | None
3816 |  Medium  | V.Likely | 326  Inadequate Encryption Strength | 3/13/2023 9:07 AM EDT | system.security.cryptography.pkcs.dll | System.Security.Cryptography.SymmetricAlgorithm OpenAlgorithm(System.Security.Cryptography.Asn1.AlgorithmIdentifierAsn): 32% | 1 Path | Open | None
3813 |   Medium | V.Likely | 326  Inadequate Encryption Strength | 3/13/2023 9:07 AM EDT | system.security.cryptography.pkcs.dll | System.Security.Cryptography.SymmetricAlgorithm OpenCipher(Asn1.AlgorithmIdentifierAsn, System.Nullable, ref System.Span /*1*/): 70% | 1 Path | Open | None
3811 |  Medium  | V.Likely | 326  Inadequate Encryption Strength | 3/13/2023 9:07 AM EDT | system.security.cryptography.pkcs.dll | System.Security.Cryptography.SymmetricAlgorithm OpenCipher(Asn1.AlgorithmIdentifierAsn, System.Nullable, ref System.Span /*1*/): 32% | 1 Path | Open | None
3806 |   Medium | V.Likely | 326  Inadequate Encryption Strength | 3/13/2023 9:07 AM EDT | system.security.cryptography.pkcs.dll | int Decrypt(ref Asn1.AlgorithmIdentifierAsn /*8*/, System.ReadOnlySpan, System.ReadOnlySpan, System.ReadOnlySpan, System.Span): 57% | 1 Path | Open | None
3808 |  Medium  | Likely | 327  Use of a Broken or Risky Cryptographic Algorithm | 3/13/2023 9:07 AM EDT | system.security.cryptography.pkcs.dll | int Decrypt(ref Asn1.AlgorithmIdentifierAsn /*8*/, System.ReadOnlySpan, System.ReadOnlySpan, System.ReadOnlySpan, System.Span): 41% | 1 Path | Open | None


</body>
</html>3821

**Comment by jstedfast** [comment]
system.security.cryptography.pkcs is part of the .NET class libraries

**Comment by bchornii** [comment]
Yes, but MimeKit 3.6.0 references System.security.cryptography.pkcs 6.0 and it seems that latest version of this library (https://www.nuget.org/packages/System.Security.Cryptography.Pkcs) is 7.0.1. Can you please update MimeKit to reference latest? Then I will be able to start discussion under appropriate repository, since usually folks advise to update to the latest and see how it goes

**Comment by bchornii** [comment]
The difference is that I will have to make a justification why I decided to install that package. Usually folks strive to have latest patches, so that's why I asked

**Comment by jstedfast** [comment]
Libraries should not do that. Only programs should.

**Comment by jstedfast** [comment]
If you test System.Security.Cryptography.Pkcs 6.0.2 (latest version that was released alongside net6.0 LTS) and it resolves any security issues, I'll update MimeKits dependency to 6.0.2
